paymentapi/src/main/java/fi/hel/verkkokauppa/payment/service/refund/PaytrailRefundPaymentService.java

```diff
@@ -0,0 +1,343 @@
+import fi.hel.verkkokauppa.payment.api.data.PaymentReturnDto;
+import fi.hel.verkkokauppa.payment.api.data.refund.RefundRequestDataDto;
+import fi.hel.verkkokauppa.payment.model.Payment;
+import fi.hel.verkkokauppa.payment.model.PaymentStatus;
+import fi.hel.verkkokauppa.payment.model.refund.RefundGateway;
+import fi.hel.verkkokauppa.payment.model.refund.RefundPayment;
+import fi.hel.verkkokauppa.payment.model.refund.RefundPaymentStatus;
+import fi.hel.verkkokauppa.payment.paytrail.PaytrailPaymentClient;
+import fi.hel.verkkokauppa.payment.paytrail.context.PaytrailPaymentContext;
+import fi.hel.verkkokauppa.payment.paytrail.context.PaytrailPaymentContextBuilder;
+import fi.hel.verkkokauppa.payment.repository.refund.RefundPaymentRepository;
+import fi.hel.verkkokauppa.payment.util.RefundUtil;
+import lombok.extern.slf4j.Slf4j;
+import org.helsinki.paytrail.model.refunds.PaytrailRefundResponse;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.sql.Timestamp;
+import java.text.SimpleDateFormat;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+@Service
+@Slf4j
+public class PaytrailRefundPaymentService {
+
+    @Autowired
+    private RefundPaymentRepository refundPaymentRepository;
+
+    @Autowired
+    private SendEventService sendEventService;
+
+    @Autowired
+    private SendNotificationService sendNotificationService;
+
+    @Autowired
+    private PaytrailPaymentContextBuilder paytrailPaymentContextBuilder;
+
+    @Autowired
+    private PaytrailPaymentClient paytrailPaymentClient;
+
+
+    public void isValidUserToCreateRefundPayment(String refundId, String userId) {
+        if (userId == null || userId.isEmpty()) {
+            log.warn("creating refund payment without user rejected, refundId: " + refundId);
+            throw new CommonApiException(
+                    HttpStatus.FORBIDDEN,
+                    new Error("rejected-creating-payment-for-order-without-user", "rejected creating refund payment for refund without user, refundId id [" + refundId + "]")
+            );
+        }
+    }
+
+    public void isValidRefundStatusToCreateRefundPayment(String refundId, String refundStatus) {
+        // check refund status, can only create refund payment for confirmed refunds
+        if (!"confirmed".equals(refundStatus)) {
+            log.warn("creating refund payment for unconfirmed refund rejected, refundId: " + refundId);
+            throw new CommonApiException(
+                    HttpStatus.FORBIDDEN,
+                    new Error("rejected-creating-refund-payment-for-unconfirmed-refund", "rejected creating refund payment for unconfirmed refund, refundId id [" + refundId + "]")
+            );
+        }
+    }
+
+    public void setPaymentStatus(String paymentId, String status) {
+        RefundPayment payment = getRefund(paymentId);
```

Comment:
Since in our context "refunds" and "refundPayments" are separate concepts, I wondered if it would be clearer to rename the method getRefund() -> getRefundPayment()?

Reply:
I am not using that function, so i just put it as placeholder for you

Reply:
Well, would it be all the same to you if the method is renamed as proposed?
Overall this is a very minor thing, I almost feel bad to even ask. :D

Reply:
Yeh go for it, you can and should rename that when you implement real functionality to that function